src/armature/Model.ts

```diff
@@ -0,0 +1,146 @@
+import { mat3, mat4 } from 'gl-matrix';
+
+import { Node } from './Node';
+import { NodeRenderObject } from './NodeRenderObject';
+
+type RenderInfo = {
+    /**
+     * The position transformation matrix for a node's coordinate space.
+     */
+    currentMatrix: mat4;
+
+    /**
+     * The normal transformation matrix for a node's coordinate space.
+     */
+    currentNormalMatrix: mat3;
+
+    /**
+     * The geometry and bones to be rendered from a node.
+     */
+    objects: NodeRenderObject;
+};
+
+/**
+ * A set of connected armature nodes, enabling efficient creation of a copy that one can add to
+ * without modifying the original. Nodes have connections to the parent that they are connected
+ * to and not to their children, so new nodes can be added that refer to parents without the parents
+ * needing to be modified.
+ */
+export class Model {
+    /**
+     * The collection of connected nodes.
+     */
+    public readonly nodes: Node[];
+
+    /**
+     * Creates a new model.
+     *
+     * @param {Node[]} nodes A set of nodes which, if passed in, are used to initialize the model.
+     */
+    constructor(nodes: Node[] = []) {
+        this.nodes = nodes;
+    }
+
+    /**
+     * Creates a new model.
+     *
+     * @param {Node[]} nodes A set of nodes which, if passed in, are used to initialize the model.
+     * @returns {Model} The new model.
+     */
+    public static create(...nodes: Node[]): Model {
+        return new Model(nodes);
+    }
+
+    /**
+     * @returns A copy of the current model that has all the same nodes, but can be added to.
+     */
+    public clone() {
+        return new Model([...this.nodes]);
+    }
+
+    /**
+     * Adds a node to the model.
+     *
+     * @param {Node} node The node to add.
+     * @returns Node The added node, for convenience.
+     */
+    public add(node: Node): Node {
+        this.nodes.push(node);
+
+        return node;
+    }
+
+    /**
+     * @returns {Node} The first node added to the model, which is therefore the root node.
+     */
+    public root(): Node {
+        return this.nodes[0];
+    }
+
+    /**
+     * @returns {Node} The most recently added node to the model.
+     */
+    public latest(): Node {
+        return this.nodes[this.nodes.length - 1];
+    }
+
+    /**
+     * Walks through all the nodes in the model, generating buffers to send to the renderer.
+     *
+     * @param {boolean} makeBones Whether or not to generate buffers for bones in addition to
+     * geometry.
+     * @returns NodeRenderObject The info needed by the renderer to visualize this model.
+     */
+    public traverse(makeBones: boolean): NodeRenderObject {
```

Comment:
This `traverse` method doesn't just _traverse_ the nodes, but it also generates buffers to send. Could we make the name a bit more descriptive, please? Like: `traverseAndSendToRenderer`.